### PR TITLE
Enhanced the Cloud Foundry (cf) client and configuration for streamlined efficiency and integration

### PR DESCRIPTION
**Pull request by ckganesan**
Enhanced the Cloud Foundry (cf) client and configuration for the following reasons:

1. To enable execution of both authenticated and non-authenticated requests using a single client, replacing the need for multiple clients.
2. To incorporate support for OAuth scopes and refresh tokens, fully integrated with context handling.
3. To extend the capabilities for reading additional information from the Cloud Foundry configuration.

### CODE REVIEW

**Comment by ckganesan** [comment]
Hi @sneal, 

Could you please review the PR?

Thanks in advance!

**Comment by sneal** [comment]
Overall I like it, especially the idiomatic Golang functional config. I'll take a closer look later this week.

**Comment by sneal** [comment]
@ckganesan Previously the client would try to re-auth with username/password or client/secret if the refresh token expired.

**Comment by ckganesan** [comment]
@sneal: Following your suggestion, I have implemented the feature to automatically generate a new token upon the expiration of the refresh token.

File: internal/http/request.go [hunk 1]
```diff
@@ -1,82 +1,68 @@
...
-	context         context.Context
+const (
+	GrantTypeNone              = ""
+	GrantTypeClientCredentials = "client_credentials"
+	GrantTypePassword          = "password"
+	DefaultRequestTimeout      = 30 * time.Second
+	DefaultUserAgent           = "Go-CF-Client/3.0"
+	MaxRedirects               = 10
+	ErrMaxRedirects            = "stopped after maximum allowed redirects"
+	DefaultClientId            = "cf"
```
Comment (sneal) [inline internal/http/request.go:22]: Should be `DefaultClientID` - Id vs ID

**Comment by sneal** [comment]
@ckganesan Thanks for adding that back in, I validated this works properly. Just a couple small issues that need addressed and I'll merge it. Thanks again!

File: client/test_runner.go [hunk 2]
```diff
@@ -31,7 +33,7 @@ func ExecuteTests(tests []RouteTest, t *testing.T) {
 				details = tt.Description + ": " + details
 			}
 
-			c, _ := config.NewToken(serverURL, "", "fake-refresh-token")
+			c, err := config.New(serverURL, config.Token("", "fake-refresh-token"))
```
[the remark this answers is not in the document]
Comment (ckganesan) [reply client/test_runner.go]: Fixed

File: internal/http/request.go (hunk 1 above)
Comment (ckganesan) [reply internal/http/request.go]: Updated

**Pull request merged**